File: bestia-client/src/Bnet/Message/Inventory/EquipItemCMSG.cs
using Bnet;

namespace BestiaBehemothClient.Bnet.Message.Inventory
{
  public partial class EquipItemCMSG : ICMSG
  {
    public ulong ItemId { get; set; }

    /// <summary>
    /// Id of the backing item instance, or 0 when the client does not know it yet.
    /// </summary>
    public ulong UniqueId { get; set; }

    /// <summary>
    /// EquipmentSlot ordinal - must match Game/Item/equipment_slot.gd (and the server enum).
    /// </summary>
    public uint Slot { get; set; }

    public override Envelope ToEnvelope()
    {
      var equipItem = new global::Bnet.EquipItemCMSG
      {
        ItemId = ItemId,
        UniqueId = UniqueId,
        Slot = Slot
      };

      return new Envelope
      {
        EquipItem = equipItem
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/item/equip/EquipItemCMSG.kt
package net.bestia.zone.item.equip

import net.bestia.bnet.proto.EquipItemCMSGProto
import net.bestia.zone.message.CMSG

data class EquipItemCMSG(
  override val playerId: Long,
  val itemId: Long,
  val uniqueId: Long,
  val slot: EquipmentSlot
) : CMSG {

  companion object {
    /**
     * Returns null for a slot ordinal the server does not know - a client sending garbage should be
     * ignored, not crash the dispatch.
     */
    fun fromBnet(playerId: Long, bnet: EquipItemCMSGProto.EquipItemCMSG): EquipItemCMSG? {
      val slot = EquipmentSlot.entries.getOrNull(bnet.slot) ?: return null

      return EquipItemCMSG(playerId, bnet.itemId, bnet.uniqueId, slot)
    }
  }
}
